text/0000-gradual-struct-init.md

```diff
@@ -0,0 +1,784 @@
+[product type]: https://en.wikipedia.org/wiki/Product_type
+[tuple type]: https://doc.rust-lang.org/nightly/reference/types.html#tuple-types
+[struct type]: https://doc.rust-lang.org/nightly/reference/types.html#struct-types
+[visibility]: https://doc.rust-lang.org/nightly/reference/visibility-and-privacy.html
+[place]: https://doc.rust-lang.org/nightly/reference/expressions.html#place-expressions-and-value-expressions
+[variable]: https://doc.rust-lang.org/nightly/reference/variables.html
+
+## Definitions, concepts, and vocabulary
+
+Let's first get some definitions out of the way; if you are familiar with the
+Rust language, you can skip this section.
+
+- A *local binding* or *[variable]* is a name for a component on the stack
+  which holds a value. An example (1):
+
+  ```rust
+  let foo = 1;
+  ```
+  
+  A binding may be *immutable*, which `foo` is, or *mutable*, which `bar` is (2):
+
+  ```rust
+  let mut foo = 1;
+  ```
+
+- An *uninhabited type* is a type which cannot be constructed since it has
+  no values that inhabit the type. An example of such a type is:
+
+  ```rust
+  enum Void {}
+  ```
+
+- *Initialization* is the process of assigning a value to a binding or more
+  generally *[place]* for the first time. Examples are in (1) and (2).
+
+- *Reinitialization* is the process of assigning a value to a
+  place that has already been initialized but then moved out of.
+  Immutable bindings may not be reinitialized.
+
+- A *[tuple type]* is denoted as `(T0, ..., Tn)`, where `Ti` are types.
+  Tuples are *[structurally typed]* homogeneous *[product type]s*.
+
+- A *[struct type]*, declared as `struct Name { field: T0, ... }` is a
+  *[nominally typed]* homogeneous product type. Structs consist of a
+  set of *fields* with types `Ti`. Given a place `x` of type `Name`
+  it is possible to project out the value of a field with `x.field`.
+
+  Each field of a struct type may optionally be assigned a *[visibility]*.
+  If no visibility is specified, it is assumed to be private by default.
+  Visibility is context dependent; a field may be visible in some context
+  but not in some other.
+
+  A struct type can also be of the tuple variety and is then
+  declared with the syntax `struct Name(T0, ..., Tn);`.
+  In that case, the field names are positional indices (`i ∈ 0 ... n`).
+
+  A struct type may also be of the unit variety, e.g. `struct Name;` in which
+  case it has no fields.
+
+## Proposal
+
+### The basics
+
+Suppose that you have a `struct` (1):
+
+```rust
+struct Foo<T> {
+    bar: usize,
+    qux: T,
+}
+```
+
+Suppose also that you have a function that consumes a `Foo<T>` (2):
+
+```rust
+fn consume<T>(_foo: Foo<T>) {}
+```
+
+Currently, if you want to make a `Foo` then you have to write (3):
+
+```rust
+let foo = Foo {
+    bar: 42usize,
+    qux: 24u8,
+};
+```
+
+**_We propose_** that you should be able to also write (4):
+
+```rust
+let foo;
+foo.bar = 42usize;
+foo.qux = 24u8;
+
+consume(foo); // OK!
+```
```

Comment:
But this RFC proposes *much more* from what I can see! I am totally onboard with fixing this (seemingly) small inconsistency (after initializing all fields, the entire type is considered initialized; I hope there is an exception for types implementing `Drop`). However, self-referential structs go way beyond this. (I will elaborate on that point elsewhere.) I find it misleading to summarize this RFC the way you do here.

Reply:
(There's an exception for types implementing `Drop` -- tho it's an unresolved q atm...)

I'll try to squeeze the into the `# Summary` about self-referential structs as well; but I wanted to have a gradual (hah!) teaching way of explaining the concepts and not throw self-referential stuff at the user directly here.

"The basics" is not the entirety of what the RFC does; just the most simple parts of it.